mlflow/models/utils.py

```diff
@@ -487,24 +532,44 @@ def _is_sparse_matrix(x):
         elif isinstance(pf_input, pd.DataFrame):
             new_pf_input = {}
             for col_name, tensor_spec in zip(input_schema.input_names(), input_schema.inputs):
-                new_pf_input[col_name] = _enforce_tensor_spec(
-                    np.array(pf_input[col_name], dtype=tensor_spec.type), tensor_spec
+                pd_series = pf_input[col_name]
+                new_pf_input[col_name] = _reshape_and_cast_pandas_column_values(
+                    col_name, pd_series, tensor_spec
                 )
         else:
             raise MlflowException(
                 "This model contains a tensor-based model signature with input names, which"
-                " suggests a dictionary input mapping input name to tensor, but an input of"
-                " type {0} was found.".format(type(pf_input))
+                " suggests a dictionary input mapping input name to tensor, or a pandas"
+                " dataframe input containing columns mapping input name to list value flattened"
+                " from tensor, but an input of type {0} was found.".format(type(pf_input))
             )
     else:
+        tensor_spec = input_schema.inputs[0]
         if isinstance(pf_input, pd.DataFrame):
-            new_pf_input = _enforce_tensor_spec(pf_input.to_numpy(), input_schema.inputs[0])
+            num_input_columns = len(pf_input.columns)
+            if num_input_columns == 0:
+                raise MlflowException("Input dataframe contains zero column.")
+            elif num_input_columns == 1:
+                new_pf_input = _reshape_and_cast_pandas_column_values(
+                    "0", pf_input[pf_input.columns[0]], tensor_spec
+                )
+            else:
+                if tensor_spec.shape != (-1, num_input_columns):
+                    raise MlflowException(
+                        "This model contains a model signature with an unnamed input, in the case "
+                        "if the input data is a pandas dataframe containing multiple columns, "
+                        "the input shape must equals to (-1, number_of_dataframe_columns), "
+                        f"but got an input dataframe of {num_input_columns} columns and the "
+                        f"input shape is {tensor_spec.shape}."
```

Comment:
```suggestion
                        f"This model's signature requires inputs of shape {tensors_spec.shape}, "
                        f" but the input dataframe has {num_input_columns} columns. The dataframe
                        f" must have {tensor_spec.shape[1]} columns."
```

Reply:
^ This message only works if we know that the tensor spec is length >= 2. What if the tensor spec length is 1?

Reply:
If the tensor spec length is 1, and the input dataframe has multiple columns, then we should raise error.

Reply:
The modified error message is also not accurate when tensor spec length > 2